src/main/java/com/google/step/snippet/servlets/FeedbackServlet.java

```diff
@@ -0,0 +1,76 @@
+package com.google.step.snippet.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/vote")
+public class FeedbackServlet extends HttpServlet {
+
+  private static final String UP = "upvote";
+  private static final String DOWN = "downvote";
+  private static final String URL = "url";
+  private static final String FEEDBACK = "feedback";
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Entity feedbackEntity = null;
+    String url = request.getParameter(URL);
+
+    if (url == null) {
+      response.sendRedirect(request.getHeader("Referer"));
+    }
+
+    /* Query for card feedback by card URL */
+    Query.FilterPredicate filter = new Query.FilterPredicate(URL, FilterOperator.EQUAL, url);
+    Query query = new Query(FEEDBACK).setFilter(filter);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    feedbackEntity = datastore.prepare(query).asSingleEntity();
+    long totalVotes = 0;
+
+    if (feedbackEntity == null) {
```

Comment:
Feels like a lot of duplicated code, and I feel like you could condense this into maybe 1/4 as many lines. The general flow I would expect is

1. If there's no entity, make one with default values.
2. Figure out from the request what the change in upvotes/downvotes will be.
3. Make the changes.
4. Find the new value of upvotes/downvotes for your total.